blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+So now during a resize, we'd get to a point where we know the window is 1,000 pixels wide. Accordingly, we'd adjust the viewport of our output surface to also be 1,000 pixels wide. But the frame that we just received is only 600 pixels wide.
+
+The [optimization from 2015](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/display.cc;l=1053-1062;drc=152a83bb9fcba4ea3a7b9f8efbe0f6204f95eadd) would then go and change the width of the root render pass to also be 1,000 pixels. But it wouldn't change what the render passes would actually draw on screen. They'd still only contain instructions to draw a picture that is 600 pixels wide.
+
+Here's what that would look like:
+
+![Visualization where the frame is smaller than the viewport](/assets/img/blog/tech-talk-window-resize-behavior/visualization-5.png)
+
+The yellow area is the area in which the render passes of the frame actually drew something. It's 600 pixels wide.
+
+However, our green viewport and our red clip rect are 1,000 pixels wide. That's the area that we show on screen. (After all, the width attribute of the root render pass claimed that it would redraw the full area of 1,000 pixels.)
+
+But because we had no draw instructions for the 400 pixels on the right, those areas didn't get updated.
+
+On the first resize, we'd show black pixels there. (That's the color that we initialize the surface with.)
+
+On subsequent resizes, those areas would show whatever was drawn to them before. We'd see stale pixels.
+
+I landed a fix for this issue in [crrev.com/c/7156576](https://crrev.com/c/7156576).
+
+The fix changes what we do when we receive a frame with a different size than our window. Instead of resizing the frame and adding gutter that contains stale pixels, we [resize our viewport](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/frame_sinks/root_compositor_frame_sink_impl.cc;l=555-556;drc=78bd63326c0a2489e6b016140aa4d84014af077a) and our clip rect.
+
+![Visualization where the clip rect and viewport size are adjusted to the frame size](/assets/img/blog/tech-talk-window-resize-behavior/visualization-6.png)
+
+We clip our surface to the size of the frame that we received. We don't show anything beyond the 600 pixels that we have draw instructions for.
+
+Voila, no more gutter, no more stale pixels!
+
+(Side note: Without `supports_viewporter`, this would be an expensive operation because it would allocate a new output surface. However, with DirectComposition, we use the "viewporter" feature. So we don't reallocate the surface when we change the viewport size. We just make a different portion of it visible. Thus, it is a cheap operation.)
+
+## Backporting the patches to Electron
+
+Once the fixes made it into Chromium, we had to pull them into Electron, too.
```

Comment:
Would be good to link to https://www.electronjs.org/docs/latest/tutorial/electron-versioning somewhere maybe